### PR TITLE
feat: graceful shutdown with signal handler

### PR DESCRIPTION
**Pull request by Gelbpunkt**
This adds a graceful shutdown handler to the server, waiting for a platform-dependent signal before terminating. On Windows, this is CTRL+C, on Unix SIGINT or SIGTERM. This is mainly to remove the dependency on dumb-init for signal forwarding in the Docker image.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -114,15 +117,35 @@ async fn main() -> Result<(), Box<dyn Error>> {
...
 
-    if let Err(why) = server.await {
+    if let Err(why) = graceful.await {
         error!("Fatal server error: {}", why);
     }
 
     Ok(())
 }
 
+async fn shutdown_signal() {
```
Comment (spring4175) [inline src/main.rs:139]: Could we make these different functions, or format in the function better? It feels hard to read.

File: src/main.rs [hunk 2]
```diff
@@ -114,15 +117,35 @@ async fn main() -> Result<(), Box<dyn Error>> {
...
+    #[cfg(unix)]
+    {
+        let mut sigint = signal(SignalKind::interrupt()).expect("failed to install SIGINT handler");
+        let mut sigterm =
+            signal(SignalKind::terminate()).expect("failed to install SIGTERM handler");
+
+        tokio::select! {
+            _ = sigint.recv() => {},
+            _ = sigterm.recv() => {},
+        };
```
Comment (spring4175) [inline src/main.rs]: This feels unfortunate. Can you think of an alternative for this? I can't off the top of my head without pulling in `futures-util`.

**Review by spring4175 — DISMISSED** [review]

File: src/main.rs (hunk 2 above)
Comment (Gelbpunkt) [reply src/main.rs]: That's not possible due to the nature of the tokio signal handling module

**Review by 7596ff — DISMISSED** [review]

**Review by vilgotf — DISMISSED** [review]

**Comment by Gelbpunkt** [comment]
rebased onto trunk

**Review by AEnterprise — APPROVED** [review]

**Review by vilgotf — APPROVED** [review]

**Review by spring4175 — APPROVED** [review]

**Pull request merged**